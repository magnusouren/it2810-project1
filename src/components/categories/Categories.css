.categories {
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  width: 60%;
  gap: 1rem;
}

h2 {
  width: 60%;
  margin: 0 auto 1rem;
  padding-bottom: 0.5rem;
  border-bottom: solid 2px #f2c2a6;
  font-family: Lato, sans-serif;
  font-weight: 500;
}

@media screen and (width <= 992px) {
  .categories {
    width: 90%;
  }

  h2 {
    width: 90%;
  }
}

@media screen and (width <= 460px) {
  .categories {
    flex-direction: column;
  }
}
