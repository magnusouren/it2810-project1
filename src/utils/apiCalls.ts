import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
