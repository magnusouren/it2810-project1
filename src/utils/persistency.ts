import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
