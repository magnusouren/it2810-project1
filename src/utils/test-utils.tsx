import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
