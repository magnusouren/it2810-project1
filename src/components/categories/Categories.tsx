import './Categories.css';

import { useQuery } from '@tanstack/react-query';
import axios from 'axios';
import { FC } from 'react';

import { CategoryType } from '../../types';
import { Category } from '../category/Category';
import { Spinner } from '../loading/Loading';

export const Categories: FC = () => {
  const { data, isLoading, isSuccess } = useQuery<CategoryType[]>(['categories'], () =>
    axios
      .get('https://www.thecocktaildb.com/api/json/v1/1/list.php?c=list')
      .then((res) => res.data.drinks.map((drink: { strCategory: string }) => drink.strCategory)),
  );

  if (isLoading) return <Spinner />;

  return (
    <>
      <h2>Categories:</h2>
      {data && isSuccess && (
        <div className='categories'>
          {data.sort().map((category: CategoryType) => (
            <Category key={category} category={category} />
          ))}
        </div>
      )}
    </>
  );
};
