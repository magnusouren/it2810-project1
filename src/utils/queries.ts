import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
