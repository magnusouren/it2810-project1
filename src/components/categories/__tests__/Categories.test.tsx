import { screen, waitFor } from '@testing-library/react';
import nock from 'nock';

import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { Categories } from '../Categories';

// vi.mock('@tanstack/react-query', async () => {
//   const axios = (await vi.importActual('@tanstack/react-query')) as { default?: undefined };
//   return {
//     ...axios,
//     useQuery: () => ({
//       data: ['Beer', 'Cocoa', 'Coffee'] as CategoryType[],
//       isLoading: false,
//       isSuccess: true,
//     }),
//   };
// });

const mockApiReturn = [
  {
    strCategory: 'Beer',
  },
  {
    strCategory: 'Cocoa',
  },
  {
    strCategory: 'Coffee',
  },
];

describe('Categories', () => {
  it('Should match snapshot', async () => {
    nock('https://www.thecocktaildb.com').get(`/api/json/v1/1/list.php?c=list`).once().reply(200, {
      drinks: mockApiReturn,
    });

    const { asFragment } = renderWithRouterAndQueryClient(<Categories />);

    await waitFor(() => {
      expect(screen.getByText('Beer')).toBeDefined();
      expect(screen.getByText('Cocoa')).toBeDefined();
      expect(screen.getByText('Coffee')).toBeDefined();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it.each(['Beer', 'Cocoa', 'Coffee'])('Should show %s', async (category) => {
    renderWithRouterAndQueryClient(<Categories />);
    expect(screen.getByText(category)).toBeDefined();
  });

  it('Should not show a category when it is not in the list', async () => {
    renderWithRouterAndQueryClient(<Categories />);
    expect(screen.queryByText('Ordinary drink')).toBeNull();
  });
});
