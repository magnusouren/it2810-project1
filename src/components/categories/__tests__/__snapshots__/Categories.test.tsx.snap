// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Categories > Should match snapshot 1`] = `
<DocumentFragment>
  <h2>
    Categories:
  </h2>
  <div
    class="categories"
  >
    <a
      class="category"
      href="/drinks"
    >
      <h3>
        Beer
      </h3>
    </a>
    <a
      class="category"
      href="/drinks"
    >
      <h3>
        Cocoa
      </h3>
    </a>
    <a
      class="category"
      href="/drinks"
    >
      <h3>
        Coffee / Tea
      </h3>
    </a>
  </div>
</DocumentFragment>
`;
